### ISSUE TITLE
SimGNN: A Neural Network Approach to Fast Graph Similarity Computation

### ISSUE THREAD

**Issue by hsanchez** [hsanchez/paper-memory#4]
## TL;DR

This paper introduces SimGNN, a novel neural network-based approach to address this classic yet challenging Graph similarity search problem, aiming to alleviate the computational burden while preserving a good performance.

### Paper Link

- [arxiv.org/abs/1808.05689](https://arxiv.org/abs/1808.05689)
- [github.com/benedekrozemberczki/SimGNN](https://github.com/benedekrozemberczki/SimGNN)

### Author/Institution

- Yunsheng Bai (University of California, Los Angeles)
- Hao Ding (Purdue University)
- Song Bian (Zhejiang University, China)
- Yizhou Sun (University of California, Los Angeles)
- Wei Wang (University of California, Los Angeles)

## Overview

This paper introduces SimGNN, a novel neural network-based approach for fast graph similarity computation. SimGNN combines two strategies: (1) a learnable embedding function that maps every graph into an embedding vector, thus providing a global summary of the embedded graph; a novel attention mechanism is adopted to highlight important nodes w.r.t. a specific similarity metric; (2) a pairwise node comparison method to supplement the graph-level embeddings with fine-grained node-level information; such information enables efficient comparison of histograms from coarse graph embeddings. 

Putting it all together, SimGNN's method for comparing graphs is quadratic on graph size. This is still among the most efficient methods for graph similarity computation. SimGNN consistently achieves the best or the second-best performance on all metrics across the three datasets. Results demonstrate that SimGNN is effective at learning a good embedding function that generalizes to unseen test graphs.

### Contributions

- Addresses the challenging classic problem of graph similarity computation by considering it as a learning problem, and proposes a neural network-based approach, called SimGNN, as the solution.
-  Proposes both a learnable embedding function that maps every graph into an embedding vector and a pairwise node comparison method for efficient comparison of embeddings' histograms.
- Conducts extensive experiments on GED, a very popular graph similarity/distance metric, on 3 real network datasets.

### Distinctions from prior works

Similar in spirit to three types of related work: 

- Network/Graph Embedding, e.g., Node-level embeddings, Graph-level embeddings. This work is among the first to apply the attention mechanism for the task of graph similarity computation. At its core, it shifts away from node-level-like prediction tasks and focuses on graph-level tasks. 

- Graph similarity computation, e.g., pairwise GED computation algorithms, graph similarity search. _No clear distinction between prior work and this work was provided.

**Note:** Authors claimed that they are the first ones to consider graph similarity computation. This is incorrect as previous work done by our lab (CSL) had considered the problem of program similarity detection using the Weisfeiler Lehman Graph Kernel on two types of graph representations. 

## Methods

SimGNN algorithm:

Given a set of graphs, SimGNN transforms the node of each graph in this set into a vector, encoding the features and structural properties around each node. Then, two strategies are to compute the similarity between graphs, one based on the interaction between two graph-level embeddings, and the other based on comparing two sets of node-level embeddings. Their combined output is then fed into a fully connected neural network to get the final similarity score.

- Graph-level embedding interaction. This strategy is based on the assumption that a good graph-level embedding can encode the structural and feature information of a graph, and by interacting the two graph-level embeddings, the similarity between two graphs can be predicted. Limitations: node-level information such as the node feature distribution and graph size may be lost by the graph-level embedding.
- Pairwise node comparison. This strategy overcomes the limitations of the previous strategy by bypassing the NTN module and using the node-level embeddings directly. Since the node-level embeddings are not normalized, the sigmoid function is applied to ensure the similarities scores are in the range of (0, 1).

## Results

SimGNN consistently achieves the best or the second-best performance on all metrics across the three datasets. Results demonstrate that SimGNN is efficient and effective at learning a good embedding function that generalizes to unseen test graphs.

Future work: (1) current model can handle graphs with node types but cannot process edge features. **Action plan:** to incorporate edge labels into our model; (2) precisions at the top k results are not preserved well,  mainly due to the skewed similarity distribution in the training dataset. **Action plan:** to explore different techniques to further boost the precisions at the top k results; and (3) Exact GED scores for large graphs cannot be computed, and thus it is unclear whether the learned model generalizes to large graphs, trained nly on the exact GEDs between small graphs. **Action plan:**  to explore how the learned model generalizes to large graphs.
